src/gutter-container.coffee

```diff
@@ -0,0 +1,95 @@
+{Emitter} = require 'event-kit'
+{Subscriber} = require 'emissary'
+Gutter = require './gutter'
+
+#remove commit
+
+# This class encapsulates the logic for adding and modifying a set of gutters.
+
+module.exports =
+class GutterContainer
+  Subscriber.includeInto(this)
+
+  # @param textEditor: The TextEditor to which this GutterContainer belongs.
+  constructor: (textEditor) ->
+    @gutters = []
+    @textEditor = textEditor
+    @emitter = new Emitter
+
+  destroy: ->
+    @gutters = null
+    @emitter.dispose()
+    @unsubscribe()
+
+  # Adds a gutter to the left side of the line numbers. Returns a Gutter instance.
+  # @param options: An object with the following fields:
+  #   name (required) A unique string to identify this gutter.
+  #   priority (optional): Determines stacking order between gutters. Lower priority items
+  #     are forced closer to the edges of the window. (default: -100)
+  #   visible (optional): boolean specifying whether the gutter is visible
+  #     initially after being created. (default: true)
+  addGutter: (options) ->
+    options = options ? {}
+    gutterName = options.name
+    if gutterName == null
+      throw new Error 'A name is required to create a gutter.'
+    existingGutter = @getExistingGutterWithName gutterName
+    if existingGutter
+      throw new Error 'Tried to create a gutter with a name that is already in use.'
+    newGutter = new Gutter this, options
+    @insertSortedByPriority newGutter, @gutters
+    newGutter
+
+  getGutters: ->
+    @gutters.slice()
+
+  ###
+  Section: Event Subscription
+  ###
+
+  # @param callback: function( removedGutter )
+  onGutterRemove: (callback) ->
+    @emitter.on 'did-remove-gutter', callback
+
+  ###
+  Section: Protected Methods
+  ###
+
+  # Processes the destruction of the gutter. Throws an error if this gutter is
+  # not within this gutterContainer.
+  removeGutter: (gutter) ->
+    index = @gutters.indexOf gutter
+    if index > -1
+      @gutters.splice(index, 1)
+      @unsubscribe gutter
+      @emitter.emit 'did-remove-gutter', gutter
+    else
+      throw new Error 'The given gutter cannot be removed because it is not ' +
+          'within this GutterContainer.'
+
+  ###
+  Section: Private Helper Methods
+  ###
+
+  getExistingGutterWithName: (name) ->
+    for gutter in @gutters
+      if gutter.name == name then return gutter
+    null
+
+  # Insert the object into the array, sorted in ascending order by the objects'
+  # 'priority' property.
+  insertSortedByPriority: (obj, array) ->
```

Comment:
Still think this could be inlined unless we want to extract a helper that is used by gutters, panels, etc.